### PR TITLE
feat() add host matching support to controllers

### PR DESCRIPTION
**Pull request by nonplus**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

## What is the current behavior?

Currently the request hostname is not taken into consideration when dispatching routes.

Issue Number: #2521

## What is the new behavior?

* New optional `host` `@Controller` setting that changes controller to only route if the request host matches.
* The `host` value may capture parameters, e.g. `:tenant.example.com`
* New `@HostParam` parameter decorator that returns captured host parameter, e.g. `@HostParam('tenant') tenant: string`
* **Caveat**: Due to how **Fastify** implements routing, this feature only works in the default (Express) router.  When used with Fastify, routing to a controller with a`{ host: ":tenant.example.com" }` filter will return a `500` response with an `HTTP Adapter does not support filtering on { host: ":tenant.example.com" }` message.

```
@Controller({ host: ':tenant.example.com')
export class TenantController {
  @Get()
  public getInfo(@HostParam('tenant') tenant: string) {
    return `Tenant from hostname: [${tenant}]`;
  }
}
```

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information

### CODE REVIEW

**Comment by kamilmysliwiec** [comment]
Thanks for your contribution! I'll review it shortly.

File: packages/core/router/router-explorer.ts [hunk 1]
```diff
@@ -185,35 +192,56 @@ export class RouterExplorer {
...
     paths.forEach(path => {
       const fullPath = stripSlash(basePath) + path;
-      routerMethod(stripSlash(fullPath) || '/', proxy);
+      routerMethod(stripSlash(fullPath) || '/', hostHandler);
     });
   }
 
+  private applyHostFilter(host, handler) {
+    if (!host) {
+      return (req, res, next) => {
```
Comment (kamilmysliwiec) [inline packages/core/router/router-explorer.ts]: Can't we just return `handler` here? (without setting the `hosts` property on `req`)

**Comment by kamilmysliwiec** [comment]
Would you like to create a PR to the docs with this feature? :)

**Comment by nonplus** [comment]
@kamilmysliwiec, the docs are done via the docs.nestjs.com repo, right? I'll open a PR today.

**Comment by nonplus** [comment]
PR for documentation update: https://github.com/nestjs/docs.nestjs.com/pull/737

**Comment by kamilmysliwiec** [comment]
Thank you! This will be published as 6.12.0

**Pull request merged**